fabric-extensibility-api-v1/src/testmod/java/net/fabricmc/fabric/test/extensibility/trident/TestTrident.java

```diff
@@ -0,0 +1,142 @@
+ * This trident allows the user to riptide from lava and burns entities when thrown at them.
+ */
+public class TestTrident extends SimpleTridentItem {
+	public TestTrident(Item.Settings settings) {
+		super(settings, new Identifier("fabric-extensibility-api-v1-testmod", "textures/entity/test_trident.png"));
+	}
+
+	@Override
+	public TridentEntity getTridentEntity(TridentEntity trident) {
+		return new SimpleTridentItemEntity(trident) {
+			@Override
+			protected void onHit(LivingEntity target) {
+				super.onHit(target);
+				target.setOnFireFor(5);
+			}
+		};
+	}
+
+	@Override
+	public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand hand) {
+		ItemStack itemStack = user.getStackInHand(hand);
+
+		if (itemStack.getDamage() >= itemStack.getMaxDamage() - 1) {
+			return TypedActionResult.fail(itemStack);
+		} else if (EnchantmentHelper.getRiptide(itemStack) > 0 && !(user.isTouchingWaterOrRain() || user.isInLava())) {
+			return TypedActionResult.fail(itemStack);
+		} else {
+			user.setCurrentHand(hand);
+			return TypedActionResult.consume(itemStack);
+		}
+	}
+
+	@Override
+	public void onStoppedUsing(ItemStack stack, World world, LivingEntity user, int remainingUseTicks) {
+		if (user instanceof PlayerEntity) {
+			PlayerEntity playerEntity = (PlayerEntity) user;
+			int i = getMaxUseTime(stack) - remainingUseTicks;
+
+			if (i >= 10) {
+				int riptideLevel = EnchantmentHelper.getRiptide(stack);
+
+				if (riptideLevel <= 0 || playerEntity.isTouchingWaterOrRain() || playerEntity.isInLava()) {
+					if (!world.isClient) {
+						stack.damage(1, playerEntity, (p) -> p.sendToolBreakStatus(user.getActiveHand()));
+
+						if (riptideLevel == 0) {
+							TridentEntity tridentEntity = new TridentEntity(world, playerEntity, stack);
+							tridentEntity.setProperties(playerEntity, playerEntity.pitch, playerEntity.yaw, 0.0F, 2.5F + riptideLevel * 0.5F, 1.0F);
+
+							if (playerEntity.abilities.creativeMode) {
+								tridentEntity.pickupType = PersistentProjectileEntity.PickupPermission.CREATIVE_ONLY;
+							}
+
+							tridentEntity = this.getTridentEntity(tridentEntity);
+							world.spawnEntity(tridentEntity);
+							world.playSoundFromEntity(null, tridentEntity, SoundEvents.ITEM_TRIDENT_THROW, SoundCategory.PLAYERS, 1.0F, 1.0F);
+
+							if (!playerEntity.abilities.creativeMode) {
+								playerEntity.inventory.removeOne(stack);
+							}
+						}
+					}
+
+					playerEntity.incrementStat(Stats.USED.getOrCreateStat(this));
+
+					if (riptideLevel > 0) {
+						float f = playerEntity.yaw;
+						float g = playerEntity.pitch;
+						float h = -MathHelper.sin(f * 0.017453292F) * MathHelper.cos(g * 0.017453292F);
+						float k = -MathHelper.sin(g * 0.017453292F);
+						float l = MathHelper.cos(f * 0.017453292F) * MathHelper.cos(g * 0.017453292F);
+						float m = MathHelper.sqrt(h * h + k * k + l * l);
+						float n = 3.0F * ((1.0F + riptideLevel) / 4.0F);
+						h *= n / m;
+						k *= n / m;
+						l *= n / m;
```

Comment:
all of these magic numbers are not really maintainable imo

Reply:
Removed from the test mod